database/engine/metadata_log/compaction.c

```diff
@@ -106,12 +110,15 @@ static void compact_record_by_uuid(struct metalog_instance *ctx, uuid_t *uuid)
         case GUID_TYPE_DIMENSION:
             rd = metalog_get_dimension_from_uuid(ctx, uuid);
             if (rd) {
+                if (ctx->current_compaction_id > rd->rrdset->rrdhost->compaction_id) {
+                    error("Forcing compaction of HOST %s", rd->rrdset->rrdhost->hostname);
+                    compact_record_by_uuid(ctx, &rd->rrdset->rrdhost->host_uuid);
+                }
                 if (ctx->current_compaction_id > rd->rrdset->compaction_id) {
-                    error("Forcing compaction of chart %s", rd->rrdset->id);
-                    rd->rrdset->compaction_id = ctx->current_compaction_id;
-                    buffer = metalog_update_chart_buffer(rd->rrdset, ctx->current_compaction_id);
-                    metalog_commit_record(ctx, buffer, METALOG_COMMIT_CREATION_RECORD, rd->rrdset->chart_uuid, 1);
-                } else if (ctx->current_compaction_id > rd->state->compaction_id) {
+                    error("Forcing compaction of CHART %s", rd->rrdset->id);
+                    compact_record_by_uuid(ctx, rd->rrdset->chart_uuid);
+                }
+                if (ctx->current_compaction_id > rd->state->compaction_id) {
```

Comment:
```suggestion
                else if (ctx->current_compaction_id > rd->state->compaction_id) {
```
This should be `else if` as before so that the dimension record is not dumped into the log twice, once from the chart dump and another one here. This is important so that the compacted log is optimal and we don't enter an infinite loop of compactions.

Reply:
True no need to do the check in that case, I will change it.

It would not be a problem with duplicate dimension records though because the `rd->state->compaction_id` would have been updated

Reply:
changed